src/Storages/MergeTree/MergeTreeDataPartDistinctPartitionExpressionCloner.cpp

```diff
@@ -0,0 +1,81 @@
+#include "MergeTreeDataPartDistinctPartitionExpressionCloner.h"
+#include <Common/escapeForFileName.h>
+#include <Interpreters/MergeTreeTransaction.h>
+
+namespace DB
+{
+
+    MergeTreeDataPartDistinctPartitionExpressionCloner::MergeTreeDataPartDistinctPartitionExpressionCloner(
+        MergeTreeData * merge_tree_data_,
+        const DataPartPtr & src_part_,
+        const MergeTreePartInfo & dst_part_info_,
+        const String & tmp_part_prefix_,
+        const MergeTreeTransactionPtr & txn_,
+        const MergeTreePartition & new_partition_,
+        const IMergeTreeDataPart::MinMaxIndex & new_min_max_index_
+    )
+    : MergeTreeDataPartCloner(merge_tree_data_, src_part_, merge_tree_data_->getInMemoryMetadataPtr(),
+                              dst_part_info_, tmp_part_prefix_, txn_, false, {}, false),
+        new_partition(new_partition_), new_min_max_index(new_min_max_index_)
+    {}
+
+    void MergeTreeDataPartDistinctPartitionExpressionCloner::deleteMinMaxFiles(
+        const DataPartStorageBuilderPtr & storage_builder
+    ) const
+    {
+        for (const auto & column : src_part->getColumns())
+        {
+            auto file = "minmax_" + escapeForFileName(column.name) + ".idx";
+            storage_builder->removeFile(file);
+        }
+    }
+
+    void MergeTreeDataPartDistinctPartitionExpressionCloner::updateMinMaxFiles(
+        const MutableDataPartPtr & dst_part,
+        const DataPartStorageBuilderPtr & storage_builder
+    ) const
+    {
+        deleteMinMaxFiles(storage_builder);
+
+        [[maybe_unused]] auto written_files = dst_part->minmax_idx->store(*merge_tree_data, storage_builder, dst_part->checksums);
```

Comment:
This looks odd (as you already mentioned), maybe add a check that `written_files` included into already existing files for the partition (before removal), and even though technically it is not an error, since this may work even for parts without those files, it will be an error after, since new files will not be added to list of files for partition (and for example checksum will not be calculated for them, so partition will be "corrupted")

Reply:
I am not sure I understand what you mean, could you clarify for me?

Reply:
At first I thought that these are the vector of paths on disk, and I suggested to ensure that dst_part already contains these paths, but I missed that `IMergeTreeDataPart::MinMaxIndex::store` already doing everything (adding checksum for those files).

And now I guess something like this can be written here:

```c
for (auto & file : written_files)
{
    file->finalize();
    if (sync)
        file->sync();
}
```

Like in other places:

https://github.com/ClickHouse/ClickHouse/blob/c8872fe96d3a6309724070b561475eec3678252e/src/Storages/MergeTree/MergedBlockOutputStream.cpp#L100-L105

Reply:
Unconditionally - definitely now, since it does fsync/fsyncdata and will slow down things (it relies on various fsync related settings)

Reply:
Ok, will leave `sync` out of the question for simplicity. If necessary, it can be added via constructor of `MergeTreeDataPartDistinctPartitionExpressionCloner`

Reply:
Done